### DISCUSSION TITLE
TypeError: Cannot read property 'INTERNAL' of undefined at FirebaseNamespace.initializeApp

### DISCUSSION THREAD

**Discussion by liamdanielduffy** [vercel/next.js#8000] [Help]
# Bug report

## Describe the bug

I've filed the issue with a reproduction [here](https://github.com/firebase/firebase-admin-node/issues/593) on the firebase-admin-node repo, but it seems to be a problem that specifically occurs when running this code in a Next 9 app, so I also want to post it here.

The firebase admin SDK initialization throws an error when I attempt to initialize it inside of a lamdba defined through the `pages/api` routing method.

Is there any special setup required for firebase-admin to be used inside of a Next app?

## To Reproduce

Visit the issue link to see the reproduction repo: https://github.com/firebase/firebase-admin-node/issues/593

**Comment by liamdanielduffy** [comment]
@Timer this actually doesn't seem to be relevant to any issue with firebase assets in a Next app.

Rather, the issue wound up being an issue with the order in which modules were imported. 

However, debugging surfaced another issue: Next.js seems to keep a Node process around when re-compiling, with an old instance of the firebase admin app in memory. This throws errors, since only one firebase admin app can be initialized at a time.

The full details of this are on this spectrum chat:

https://spectrum.chat/firebase/general/typeerror-cannot-read-property-internal-of-undefined-at-firebasenamespace-initializeapp~a839b2c8-2349-4bf9-9b21-b4163a916119

**Comment by Timer** [comment]
Interesting bug! Sorry about this, but it's likely due to our HMR reloading of the API endpoint.

We have an open issue (#7874) to ensure these endpoints have Lambda-like (Run-to-Completion) semantics that should fix this.

**Comment by liamdanielduffy** [comment]
Awesome 🙂 thank you!

**Comment by xndyz** [comment]
I am getting this error as well with Firebase when importing a JSON key.

```
nextjs            | TypeError: Cannot read property 'INTERNAL' of undefined
nextjs            |     at FirebaseNamespace.initializeApp (/app/node_modules/firebase-admin/lib/firebase-namespace.js:409:21)
nextjs            |     at Module../pages/api/account/create.ts (/app/.next/server/static/development/pages/api/account/create.js:112:71)
nextjs            |     at __webpack_require__ (/app/.next/server/static/development/pages/api/account/create.js:23:31)
nextjs            |     at Object.5 (/app/.next/server/static/development/pages/api/account/create.js:161:18)
nextjs            |     at __webpack_require__ (/app/.next/server/static/development/pages/api/account/create.js:23:31)
nextjs            |     at /app/.next/server/static/development/pages/api/account/create.js:91:18
nextjs            |     at Object.<anonymous> (/app/.next/server/static/development/pages/api/account/create.js:94:10)
nextjs            |     at Module._compile (internal/modules/cjs/loader.js:1158:30)
nextjs            |     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1178:10)
nextjs            |     at Module.load (internal/modules/cjs/loader.js:1002:32)
nextjs            |     at Function.Module._load (internal/modules/cjs/loader.js:901:14)
nextjs            |     at Module.require (internal/modules/cjs/loader.js:1044:19)
nextjs            |     at require (internal/modules/cjs/helpers.js:77:18)
nextjs            |     at DevServer.handleApiRequest (/app/node_modules/next/dist/next-server/server/next-server.js:430:28)
nextjs            |     at async Object.fn (/app/node_modules/next/dist/next-server/server/next-server.js:361:37)
nextjs            |     at async Router.execute (/app/node_modules/next/dist/next-server/server/router.js:137:32)
```

**Comment by kalwabed** [comment]
Have you tried the solution from https://github.com/firebase/firebase-admin-node/issues/593#issuecomment-908616694. I faced the same issue, and it was fixed when I used `import * as admin from 'firebase-admin'` as mentioned in that comment.
